storage/engine/mvcc.go

```diff
@@ -423,57 +425,64 @@ type getValueFunc func(engine Engine, start, end proto.EncodedKey,
-			return nil, err
+			return nil, nil, err
 		}
-		return meta.Value, nil
-	}
-	// If we're doing inconsistent reads and there's an intent, we
-	// ignore the intent by insisting that the timestamp we're reading
-	// at is a historical timestamp < the intent timestamp. However, we
-	// return a write intent error so that the intents can be resolved.
-	var wiErr error
-	if !consistent && meta.Txn != nil && !timestamp.Less(meta.Timestamp) {
+		return meta.Value, nil, nil
+	}
+	var ignoredIntents []proto.Intent
+	if meta.Txn != nil && !timestamp.Less(meta.Timestamp) && !consistent {
+		// If we're doing inconsistent reads and there's an intent, we
+		// ignore the intent by insisting that the timestamp we're reading
+		// at is a historical timestamp < the intent timestamp. However, we
+		// return the intent separately; the caller may want to resolve it.
+		ignoredIntents = append(ignoredIntents, proto.Intent{Key: key, Txn: *meta.Txn})
 		timestamp = meta.Timestamp.Prev()
-		wiErr = &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
 	}
 
 	var valueKey proto.EncodedKey
 	value := &buf.value
 
+	var err error
 	// First case: Our read timestamp is ahead of the latest write, or the
-	// latest write and current read are within the same transaction.
-	if !timestamp.Less(meta.Timestamp) ||
-		(meta.Txn != nil && txn != nil && bytes.Equal(meta.Txn.ID, txn.ID)) {
-		if meta.Txn != nil && (txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID)) {
-			// Trying to read the last value, but it's another transaction's
-			// intent; the reader will have to act on this.
-			return nil, &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
+	// latest write and current read are within the same (non-nil) transaction.
+	if !timestamp.Less(meta.Timestamp) && meta.HasWriteIntentError(txn) {
+		// Trying to read the last value, but it's another transaction's
+		// intent; the reader will have to act on this.
+		return nil, nil, &proto.WriteIntentError{
+			Intents: []proto.Intent{{Key: key, Txn: *meta.Txn}},
 		}
+		// TODO remove
+	} else if ownIntent := meta.IsIntentOf(txn); !timestamp.Less(meta.Timestamp) || ownIntent {
+		// We want the latest value, and it's not an intent or our intent, both
+		// of which are ok. Note that when reading the own intent, the timestamp
+		// specified is irrelevant; we always want to see the intent (see
+		// TestMVCCReadWithPushedTimestamp).
 		latestKey := mvccEncodeTimestamp(metaKey, meta.Timestamp)
 
 		// Check for case where we're reading our own txn's intent
 		// but it's got a different epoch. This can happen if the
 		// txn was restarted and an earlier iteration wrote the value
 		// we're now reading. In this case, we skip the intent.
-		if meta.Txn != nil && txn.Epoch != meta.Txn.Epoch {
+		if ownIntent && txn.Epoch != meta.Txn.Epoch {
 			valueKey, err = getValue(engine, latestKey.Next(), MVCCEncodeKey(key.Next()), value)
```

Comment:
I removed the function-scope err variable and added them in each of the cases. For this particular one it's now sharing one with the `else {...}` clause.